Create comment when linking to existing JIRA issue

earshinov opened the pull request:
Fixes #21865

![image](https://user-images.githubusercontent.com/52021/104333184-02c9b080-5502-11eb-83aa-86d116c5f388.png)

A brief feature description:
  - Users get a pre-populated Description field on the Link tab
  - The default description is what you would have in a newly created issue, but starting with "Linked Sentry Issue" instead of "Sentry Issue". If you don't think this change is necessary, I will gladly remove it.
  - On the backend, when the JIRA integration receives a non-empty description field, it creates a comment in the linked JIRA ticket.
  -  JIRA and JIRA Server integrations are affected.

Please have mercy on me and don't ask me to implement additional features like adding a comment when JIRA issue is unlinked. Setting up development environment in WSL (and development itself) is a torture, and I have spend at least 3 days on this trivial feature. Assuming that it would take ~10x less time for Sentry developers to implement this themselves, I don't think my time is so dirt cheap. (Although, I must admit that this torture was somewhat educational).

I am ready to add tests if required.

earshinov commented:
FYI, failure of at least some of the checks is unrelated to my changes. For example (https://github.com/getsentry/sentry/runs/1689026191?check_suite_focus=true):

```
Obtaining file:///home/runner/work/sentry/sentry
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: /opt/hostedtoolcache/Python/3.6.10/x64/bin/python /opt/hostedtoolcache/Python/3.6.10/x64/lib/python3.6/site-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-y5b1wqxw/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.2.0' wheel
       cwd: None
  Complete output (2 lines):
  ERROR: Could not find a version that satisfies the requirement setuptools>=40.2.0
  ERROR: No matching distribution found for setuptools>=40.2.0
  ----------------------------------------
ERROR: Command errored out with exit status 1: /opt/hostedtoolcache/Python/3.6.10/x64/bin/python /opt/hostedtoolcache/Python/3.6.10/x64/lib/python3.6/site-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-y5b1wqxw/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.2.0' wheel Check the logs for full command output.
```

earshinov commented:
Formatted changed files with `black` and force-puhed.

manuzope commented:
Hey @earshinov, thanks for creating the PR!

Would it be possible to remove this from JIRA and only keep it for JIRA server? We have [issue glances](https://developer.atlassian.com/cloud/jira/platform/modules/issue-glance/) as a part of the JIRA integration that serves this purpose. If we can have JIRA server also have an issue glance view that would be even better.

earshinov commented:
@manuzope , hi! Thanks for your feedback. I will update my PR in the next few days so that it only affects Jira Server. I am not sure if Jira Server is going to have the same "issue glances" as the cloud version -- they seem to only mention them in the context of something called HipChat: https://developer.atlassian.com/server/hipchat/glances/.

earshinov commented:
Rebased on the latest `master` to resolve conflicts.

earshinov commented:
@manuzope , Done in e776303

ArcticSnowman commented:
Looking forward to this..   our teams really want to be able to link to existing ticket and see references to the sentry.io in the tickets.

ceorourke commented:
Hi @earshinov thanks for the contribution! After discussing this internally and taking another look at the original issue, we'd like this to only create a comment with the **link** and not include the entire description. Can you add a referrer to the link as well? Once that's ready I'll pull it down and test locally.

earshinov commented:
Hi @ceorourke ,

I have pushed an updated version.

> After discussing this internally and taking another look at the original issue, we'd like this to only create a comment with the link and not include the entire description.

I assume it is still desirable to make the default comment editable by user?

![image](https://user-images.githubusercontent.com/52021/107425153-ac1abb00-6b2f-11eb-8e5a-07842d982960.png)

> Can you add a referrer to the link as well?

Added `referrer=jira_server_integration`.

The posted Jira comment looks like this:

![image](https://user-images.githubusercontent.com/52021/107425236-c6549900-6b2f-11eb-9a9a-6ab99fa431f3.png)

ceorourke commented on src/sentry/integrations/jira_server/integration.py:
```diff
@@ -244,11 +246,35 @@ def get_link_issue_config(self, group, **kwargs):
+
+        default_comment = "Linked Sentry Issue: [{}|{}]".format(
+            group.qualified_short_id,
+            absolute_uri(group.get_absolute_url(params={"referrer": "jira_server_integration"})),
```
@earshinov I know I didn't give any guidance but can you change the referrer to just `jira_server`?

earshinov replied:
Sure, fixed in 0ba1598ae556b1636aaef0d1c17c17f330af0f4b

ceorourke commented:
/gcbrun

ceorourke commented:
@earshinov you might have to rebase to resolve the backend lint failure

earshinov commented:
@ceorourke , It seems that `backend lint` still fails for some reason.

I have squashed my commits into one and rebased. Squashing was necessary because my initial commits (which touched the base `JiraIntegration` class) caused conflicts.

ceorourke commented:
@earshinov can you run `pre-commit run pyupgrade --all-files` locally? This should help with the backend lint failure.

earshinov commented:
@ceorourke , Thanks. It seems that it fixed the problem indeed.

ceorourke commented:
/gcbrun

ceorourke commented:
@earshinov thanks again for the PR!
